pkg/controller/controller_suite_test.go

```diff
@@ -27,3 +40,215 @@ func TestMachineControllerManagerSuite(t *testing.T) {
+	machineSets := make([]*v1alpha1.MachineSet, replicas)
+	for i := range machineSets {
+		ms := &v1alpha1.MachineSet{
+			ObjectMeta: *newObjectMeta(&specTemplate.ObjectMeta, i),
+			Spec: v1alpha1.MachineSetSpec{
+				MachineClass:    *specTemplate.Spec.Class.DeepCopy(),
+				MinReadySeconds: minReadySeconds,
+				Replicas:        replicas,
+				Selector: &metav1.LabelSelector{
+					MatchLabels: deepCopy(specTemplate.ObjectMeta.Labels),
+				},
+				Template: *specTemplate.DeepCopy(),
+			},
+		}
+		if statusTemplate != nil {
+			ms.Status = *statusTemplate.DeepCopy()
+		}
+		machineSets[i] = ms
+	}
+	return machineSets
+}
+
+func deepCopy(m map[string]string) map[string]string {
+	r := make(map[string]string, len(m))
+	for k := range m {
+		r[k] = m[k]
+	}
+	return r
+}
+
+func newMachineFromMachineSet(machineSet *v1alpha1.MachineSet, statusTemplate *v1alpha1.MachineStatus) *v1alpha1.Machine {
+	return newMachinesFromMachineSet(1, machineSet, statusTemplate)[0]
+}
+
+func newMachinesFromMachineSet(machineCount int, machineSet *v1alpha1.MachineSet, statusTemplate *v1alpha1.MachineStatus) []*v1alpha1.Machine {
+	t := machineSet.TypeMeta
+	return newMachines(machineCount, &machineSet.Spec.Template, statusTemplate, &metav1.OwnerReference{
+		APIVersion:         t.APIVersion,
+		Kind:               t.Kind,
+		Name:               machineSet.Name,
+		UID:                machineSet.UID,
+		BlockOwnerDeletion: boolPtr(true),
+		Controller:         boolPtr(true),
+	})
+}
+
+func newMachine(specTemplate *v1alpha1.MachineTemplateSpec, statusTemplate *v1alpha1.MachineStatus, owner *metav1.OwnerReference) *v1alpha1.Machine {
+	return newMachines(1, specTemplate, statusTemplate, owner)[0]
+}
+
+func newMachines(machineCount int, specTemplate *v1alpha1.MachineTemplateSpec, statusTemplate *v1alpha1.MachineStatus, owner *metav1.OwnerReference) []*v1alpha1.Machine {
+	machines := make([]*v1alpha1.Machine, machineCount)
+	for i := range machines {
+		m := &v1alpha1.Machine{
+			ObjectMeta: *newObjectMeta(&specTemplate.ObjectMeta, i),
+			Spec:       *newMachineSpec(&specTemplate.Spec, i),
+		}
+
+		if statusTemplate != nil {
+			m.Status = *newMachineStatus(statusTemplate, i)
+		}
+
+		if owner != nil {
+			m.OwnerReferences = append(m.OwnerReferences, *owner.DeepCopy())
+		}
+
+		machines[i] = m
+	}
+	return machines
+}
+
+func boolPtr(b bool) *bool {
+	return &b
+}
+
+func newObjectMeta(meta *metav1.ObjectMeta, index int) *metav1.ObjectMeta {
+	r := meta.DeepCopy()
+
+	if r.Name != "" {
+		return r
+	}
+
+	if r.GenerateName != "" {
+		r.Name = fmt.Sprintf("%s-%d", r.GenerateName, index)
+		return r
+	}
+
+	r.Name = fmt.Sprintf("machine-%d", index)
+	return r
+}
+
+func newMachineSpec(specTemplate *v1alpha1.MachineSpec, index int) *v1alpha1.MachineSpec {
+	r := specTemplate.DeepCopy()
+	if r.ProviderID == "" {
```

Comment:
Should the condition be r.ProviderID != ""

Reply:
No. The during create the MachineSpec is expected to have ProviderID = nil. ProviderID is filled only for machines where the VM has been created.

So, in this case, I wanted to be able to create many machines (new or old) based on the template. If no ProviderID is given then it is expected that it is for a new machine. If ProviderID is given then it is used as a prefix.